### PR TITLE
UTF-8 language translation supports

### PR DESCRIPTION
**Pull request by yhfudev**
It supports UTF-8 language translation header files, and generate the font data automatically.
So the language maintainer can use Unicode to translate their languages and don't need care about
the font if the font contains all of the glyph of their languages.

The HD44780 is also works.

The document(howto) for the new language engine is in the file buildroot/share/fonts/uxggenpages.md.

See the discussions:
https://github.com/MarlinFirmware/Marlin/issues/4539#issuecomment-242048012

### CODE REVIEW

**Comment by yhfudev** [comment]
It needs complete U8GLIB including c code (u8g.h)!

File: Marlin/ultralcd_impl_DOGM.h [hunk 1]
```diff
@@ -620,18 +609,22 @@ static void lcd_implementation_status_screen() {
       uint8_t n = LCD_WIDTH - (START_COL) - 1;
 
       if (longFilename[0]) {
-        filename = longFilename;
+        filename = longFilename
```
Comment (thinkyhead) [inline Marlin/ultralcd_impl_DOGM.h]: Oops!

File: Marlin/ultralcd_impl_DOGM.h [hunk 2]
```diff
@@ -598,16 +587,16 @@ static void lcd_implementation_status_screen() {
...
     const float kHalfChar = (DOG_CHAR_HEIGHT_EDIT) / 2;
     float rowHeight = u8g.getHeight() / (rows + 1); // 1/(rows+1) = 1/2 or 1/3
 
-    u8g.setPrintPos(0, rowHeight + kHalfChar);
+    lcd_moveto (0, rowHeight + kHalfChar);
     lcd_printPGM(pstr);
     if (value != NULL) {
       lcd_print(':');
-      u8g.setPrintPos((lcd_width - 1 - vallen) * char_width, rows * rowHeight + kHalfChar);
+      lcd_moveto ((lcd_width - 1 - vallen) * char_width, rows * rowHeight + kHalfChar);
```
Comment (thinkyhead) [inline Marlin/ultralcd_impl_DOGM.h]: As a matter of code style, please remove spaces between the function name and the opening parenthesis.

File: Marlin/ultralcd_impl_DOGM.h [hunk 3]
```diff
@@ -587,7 +575,8 @@ static void lcd_implementation_status_screen() {
     uint8_t vallen = lcd_strlen(value);
 
     #if ENABLED(USE_BIG_EDIT_FONT)
-      if (lcd_strlen_P(pstr) <= LCD_WIDTH_EDIT - 1) {
+      //if (lcd_strlen_P(pstr) <= LCD_WIDTH_EDIT - 1) {
```
Comment (thinkyhead) [inline Marlin/ultralcd_impl_DOGM.h]: If replacing a line, you can just remove it instead of leaving it commented out.

File: Marlin/ultralcd_impl_DOGM.h [hunk 4]
```diff
@@ -522,13 +504,16 @@ static void lcd_implementation_status_screen() {
...
+#if 1
+      lcd_printstr_P (pstr, PIXEL_LEN_NOLIMIT);
+#else
     while (c = pgm_read_byte(pstr)) {
       n -= lcd_print(c);
       pstr++;
     }
     while (n--) lcd_print(' ');
-    u8g.setPrintPos(LCD_PIXEL_WIDTH - (DOG_CHAR_WIDTH), (row + 1) * (DOG_CHAR_HEIGHT));
+#endif
```
Comment (thinkyhead) [inline Marlin/ultralcd_impl_DOGM.h]: If replacing lines of code, you should just replace them. Don't bother with these faux conditional blocks.

File: Marlin/u8g_fontutf8.c [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+}
+
+typedef struct _font_group_t {
+    uxg_fontinfo_t * m_fntifo;
+    int m_fntinfo_num;
+} font_group_t;
+
+static int
+fontgroup_init (font_group_t * root, uxg_fontinfo_t * fntinfo, int number)
+{
```
Comment (thinkyhead) [inline Marlin/u8g_fontutf8.c:86]: Code in Marlin should follow the style guidelines described in the (soon to be obsolete) wiki pages. Opening parentheses go at the end of the line. Return types belong on the same line as the function name. Indentation must be 2 spaces, not 3. Etc.

File: Marlin/lcdprint_u8g.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    return ret;
+}
+
+int
+lcd_printstr (const char * utf8_str, pixel_len_t max_length)
+{
+    unsigned int ret;
+    unsigned int x;
+    unsigned int y;
```
Comment (thinkyhead) [inline Marlin/lcdprint_u8g.cpp:69]: Types that explicitly specify the bit-length are preferred. `uint16_t` instead of `unsigned int`, etc.

File: Marlin/lcdprint_hd44780.cpp [hunk 7]
```diff
@@ -0,0 +1,812 @@
...
+    {IV('⌡'), 0xC1, 0},
+
+    {IV('⎧'), 0x14, 0},
+    {IV('⎩'), 0x15, 0},
+    {IV('⎫'), 0x16, 0},
+    {IV('⎭'), 0x17, 0},
+    {IV('⎰'), 0x18, 0},
+    {IV('⎱'), 0x19, 0},
+    {IV('⎲'), 0x12, 0},
+    {IV('⎳'), 0x13, 0},
```
Comment (thinkyhead) [inline Marlin/lcdprint_hd44780.cpp:479]: That is a very cool two-part sigma!

File: Marlin/language_test.h [hunk 8]
```diff
@@ -69,39 +69,39 @@
...
-#define STRG_E382_a "゠ァアィイゥウェエォオカガキギク"
-#define STRG_E382_b "グケゲコゴサザシジスズセゼソゾタ"
-#define STRG_E383_8 "トチヂッツヅテデトドナニヌネノハ"
-#define STRG_E383_9 "バパヒビピフブプヘベペホボポマミ"
-#define STRG_E383_a "ムメモャヤュユョヨラリルレロヮワ"
-#define STRG_E383_b "ヰヱヲンヴヵヶヷヸヹヺ・ーヽヾヿ"
+#define STRG_ASCII_2 _UxGT(" !\"#$%&'()*+,-./")
+#define STRG_ASCII_3 _UxGT("0123456789:;<=>?")
+#define STRG_ASCII_4 _UxGT("@ABCDEFGHIJKLMNO")
+#define STRG_ASCII_5 _UxGT("PQRSTUVWXYZ[\]^_")
```
Comment (thinkyhead) [inline Marlin/language_test.h]: You may need to use double-backslash to get a backslash (but I'm not sure with C strings).

File: Marlin/language_pl.h [hunk 9]
```diff
@@ -32,136 +32,136 @@
...
+#define MSG_ERR_MINTEMP_BED                 _UxGT("Err: min. temp. loza")
+#define MSG_HEATING                         _UxGT("Rozgrzewanie...")
+#define MSG_HEATING_COMPLETE                _UxGT("Rozgrzano")
+#define MSG_BED_HEATING                     _UxGT("Rozgrzewanie loza...")
+#define MSG_BED_DONE                        _UxGT("Rozgrzano loze")
+#define MSG_DELTA_CALIBRATE                 _UxGT("Kalibrowanie Delty")
+#define MSG_DELTA_CALIBRATE_X               _UxGT("Kalibruj X")
+#define MSG_DELTA_CALIBRATE_Y               _UxGT("Kalibruj Y")
+#define MSG_DELTA_CALIBRATE_Z               _UxGT("Kalibruj Z")
+#define MSG_DELTA_CALIBRATE_CENTER          _UxGT("Kalibruj środek")
```
Comment (thinkyhead) [inline Marlin/language_pl.h:164]: Is `_UxGT()` really needed for all languages, even those that just use Roman and accented characters? I was under the impression these would only need to be applied where the unreadable character codes were being used.

**Comment by thinkyhead** [comment]
Overall an extraordinary effort. Well done! I hope we can get consensus among the project members on this soon. Can you supply some documentation on this for the Marlin website also?

File: Marlin/language_pl.h (hunk 9 above)
Comment (yhfudev) [reply Marlin/language_pl.h:164]: _UxGT() is a macro to be used by the script to find the UTF-8 strings. You need to use _UxGT() for all of the UTF-8 strings if you want the Marlin to support UTF-8 char in other source code, and use the script to generate the font data.

And in the future, for example, if we want to use gettext() similar solution, which use standalone translation files in SD card (font is also in SD card), we can re-define the macro _UxGT to support it.

File: Marlin/u8g_fontutf8.c (hunk 5 above)
Comment (yhfudev) [reply Marlin/u8g_fontutf8.c:86]: Do you have a "indent" settings(BASH command indent) or other tools to re-format the source code? It would be hard to change the format by manual, and may make mistakes during the formatting.

File: Marlin/ultralcd_impl_DOGM.h (hunk 1 above)
Comment (yhfudev) [reply Marlin/ultralcd_impl_DOGM.h]: Sorry, I'll submit the fix.

File: Marlin/lcdprint_u8g.cpp (hunk 6 above)
Comment (yhfudev) [reply Marlin/lcdprint_u8g.cpp:69]: Do you mean the type pixel_len_t? It's defined as uint16_t. Actually, I don't know what the proper size for this argument, I leave in the header file so others can give a better size.

Comment (thinkyhead) [reply Marlin/lcdprint_u8g.cpp:69]: When I post a line note it pertains the line it is posted on. So here I'm referring to the declarations of `ret`, `x`, and `y`.

File: Marlin/u8g_fontutf8.c (hunk 5 above)
Comment (thinkyhead) [reply Marlin/u8g_fontutf8.c:86]: No, we cannot use automated formatting. It ruins the source, which is –yes– hand coded to be readable by humans. The human who did most of the cleaning up by hand was me. Use your editor's search-and-replace, as I did — and all the better if your editor has regex support, which I constantly find to be an incredible time-saver.

**Comment by thinkyhead** [comment]
Read the output of Travis CI to see where it is hitting errors in the code, and fix them.

**Comment by yhfudev** [comment]
(Never mind, I get it re-installed by down-grade and upgraded.)

@thinkyhead,
Do you know where's the U8glib installed, after we execute

``` sh
arduino --install-library "U8glib"
```

I need to check the files in the lib files, but I can't find it in all of "libraries" folders, including $HOME/Arduino/libraries, /usr/share/Arduino/libraries etc.

And how to uninstall it. How to find the config file for the libraries?

Thanks!

**Comment by petrzjunior** [comment]
@yhfudev This feature looks amazing! I was thinking about something similar, but ended up at U8glib's chartable support. But your solution is so... great!

I just wanted to ask: If I write some non-english characters (e.g. áéěčšů) in  language file, will they display normally on HD4470 displays? And also you're talking about some `./genallfont.sh` script but I can't find it, neither in your fork, nor in wqy-bitmapfont-bdf.

**Comment by yhfudev** [comment]
It depends on hd44780 hardware(the internal glyph data). There're three or
more different type glyph ROMs: Japanese, Russian, Western. I think you
want the Western version, which I can't find in eBay. My version is
Japanese, and still waiting for Russian version arrived. You need to find
the data sheet of the module, and modify lcdprint_hd44780.cpp accordingly.

In a short, you may check the chars in the "WESTERN" block in the file lcdprint_hd44780.cpp
to see if there exists the char you want.

The genallfont.sh is in folder Marlin. And the wqy font should be in the
debian/Ubuntu package xfont-wqy. You may take a look the doc
buildroot/share/fonts/uxggenpages.md for details.

**Comment by yhfudev** [comment]
Can we merge the 9 glyph in the font Marlin_symbols to a fallback font ISO10646_1?
Since 9 is a small number, and this can simplify the implementation of the new lcd_print(), and we can also remove the lines setFont(Marlin_symbols).

**Comment by AnHardt** [comment]
Finally i'v found some time to test.

Compiling this PR is difficult (maybe too difficult for the average user).
The need to modify u8glib by copying u8g.h from .../src/clib/u8g.h to .../src/u8g.h in the library folder needs a very clear hint. I don't expect any user to find out by himself the trick. Nobody else looks into .travis.jml to see how it is made there to make it work.

Til now the optical result on 12854 displays is poor. - at least in English and German.
picture 5
All text is one pixel row too hight. Likely the baseline of the font is shifted.

picture 6
German Umlauts are extemly ugly, are not consistent with the related gyphes. 'ä' is a 'a' with double dots above. 'ü' is 'u' with double dots above. That is likely to be a error in the WQY-font.

Updating a language under WIN7 seems to be impossible. (Cygwin ?)
Updating a language under WIM10 is difficult to set up without a good step by step guide.

Install the new WIN10-Ubuntu subsystem.
Install make, gcc, pcf2bdf, xfonts-wqy, ... .
compile bdf2u8g, genpages, ... .
Convert line-breaks of .sh-scripts from /r/n to /n.
Run genallfont.sh

I got it to work - somehow - but compared with the current system, it was a nightmare.

**Comment by yhfudev** [comment]
@AnHardt
Thanks for your time testing the code!

The main problem is the font as I mentioned in the https://github.com/MarlinFirmware/Marlin/issues/4539#issuecomment-242048012;
There's no complete font for multi-language and it may not look beautiful. I didn't compile the font with Marlin's font because I'm not sure the compatibility of the font licenses. The WQY's font is only for test purpose.

For the U8GLIB header problem, we can ask the author to update his lib to expose the u8g.h file to reduce the user config time. 

The "nightmare" is not a problem since it works and you can consider it needs run only "once" of the script. In the previous implementation, the user also have to compile the bdf2u8g and run some script to generate the header file if he changes the language file or add an new language. Not to mention that he may have to select glyph from thousands by manual as an extra work if he use CJK charset, which can be done automatically in the new solution.

There still has solution to fix the minor problem if the developer feels "pain" about the script. He can either rewrites the genfont util in C to replace the script version. This is equivalence to previous version for the end user(or language maintainer) because he also needs to compile the bdf2u8g. Or he can rewrite the script by other programming language, such as python, to get a bit more portability, but it needs the end user to install python (in Windows).

In summary, the new solution is more straightforward than the previous one as it doesn't need maintain mysterious code in language file (such as "\xd2\xb2\xdd", "\xa4\xa5\xa6\xa7" etc), and both the language files and font can be easily maintained separately. And it also keeps the font data small by converting only the used chars in the language files.

**Comment by AnHardt** [comment]
> The main problem is the font as I mentioned in the #4539 (comment);
> There's no complete font for multi-language and it may not look beautiful. I didn't compile the font with Marlin's font because I'm not sure the compatibility of the font licenses. The WQY's font is only for test purpose.

Indeed. I know why my (Marlins) fonts are hand drawn. The licence of Marlins fonts is "public domain"  = "do with them whatever you want". I don't see a way to make them more 'free'.

> The "nightmare" is not a problem since it works and you can consider it needs run only "once" of the script. 

The nightmare is a problem. I'm a experienced programmer, but our target audience is not. As far i can see, the assumption this process is rare, is wrong. For example in the 'current' German translation we do not use the common symbols "ÄÜöÖß". But that is completely random. Likely tomorrow we want to use this symbols. With your system we have to rebuild the fonts, while in the old system we can simlpy use them, because they are always in the font. 
Maybe a dummy string, containing the 'common' symbols of a language can help to reduce the number of font generations.

> Not to mention that he may have to select glyph from thousands by manual as an extra work if he use CJK charset, which can be done automatically in the new solution.

Indeed. I see the progress for the CK languages. But also (currently) i see a regression for all the others.

> In summary, the new solution is more straightforward than the previous one as it doesn't need maintain mysterious code in language file (such as "\xd2\xb2\xdd", "\xa4\xa5\xa6\xa7" etc), and both the language files and font can be easily maintained separately.

I see it exactly the other way. The 'old' system is much more convenient - except for C&K. Because all needed symbols are in. You really never have to touch the fonts again when produced once.

How about a script only producing our language and font files from a utf8-template for Chinese and Korean?

A big help for the producers of printers could be, if we could select the display language at runtime from a set of n.

**Comment by AnHardt** [comment]
How do you want to handle M117 commands where 'common' symbols are used that are not in the 'compiled' text?

**Comment by yhfudev** [comment]
@AnHardt 

> "Maybe a dummy string, containing the 'common' symbols of a language can help to reduce the number of font generations."

You already answer your question, right?

> "I see it exactly the other way. The 'old' system is much more convenient - except for C&K. Because all needed symbols are in. You really never have to touch the fonts again when produced once."

So how can a maintainer to add new language support? Does he need to change the file utf_mapper.h? Is the code cover all of the (language) cases, or can it handles 3 bytes utf8 code? Does he need to copy a exist function to a new location and change some HEX values to support the language? Does he need to map the values to the font data, does he need to add new MACROS to claim the char-set? And add it to everywhere it need to add? 

The 'new' system gives the user just two interfaces to handle the language problem: the Unicode translation header and the standard Unicode BDF font file. For the questions you mentioned,
- It needs only to update the language file or add a new language file, and ask the end user to declare which language should be used in the Configuration.h by macro LCD_LANGUAGE.
- There's no need to use two separate scripts to handle different languages, it's just simple one. The problem you mentioned is equivalence to how to create a new Unicode bitmap font and adjust the font parameter. In another word, it's the problem of data, and is easier to be maintained than source code.
- For the language with small char-set, and to reduce the number of font generations, you already have a method for the 'new' system above, which is also used in the language_jp_kana.h.
- Is there road-map to support run-time multi-language? This can be supported by adding a string function base on the 'new' system.
- For M177, there's another question related to this one. SD is supported by Marlin, so how can Marlin display the file name of the locale language?
  Ok, back to the M177 question, it depends on the font it's compiled. For the German, you already have the solution above for the 'new' system; but for CJK, it's a real nightmare because of the font data size. it's also a big nightmare for Marlin showing CJK file name in FAT SD card (it needs a map table between Unicode and big5, gb2312,jisx0208,korean etc as well as font data) in either 'old' or 'new' system.
- One possible solution for M177 and also for run-time multi-language, is what I mentioned in above and https://github.com/MarlinFirmware/Marlin/issues/4539#issuecomment-242048012, using translation file and font files storing in SD card and using gettext() like function to translate the strings. We need to evaluate if this has performance penalties.

**Comment by thinkyhead** [comment]
The main question for me is: What is required to make an update to a single string an existing language. Can we just update the language file and that's it?

Second to that is, does this add more overhead to drawing strings, especially on graphical displays? And if so, how much additional overhead?

**Comment by yhfudev** [comment]
@thinkyhead
1. There's nothing to do after you update a string IF there's glyph in the existing font data. One of the scenario is the using "dummy string, containing the 'common' symbols of a language" as AnHardt mentioned. So you can "just update the language file and that's it".

Or you have to run the bash script(or C or python variations) to generate the font data to be compiled in the firmware. This is for the scenario that there's no such glyph in the font data (the char set is huge, for example # > 1000).
1. I don't think it is worse than that of one big font. U8GLIB's font supports variant width of the glyph, so the draw function have to calculate the position of the glyph data before getting the bitmap. it needs parse all of the width values before that glyph.

The 'new' system use separate fonts to store the glyph data due to the chars values are not continued.  
The font data are sorted and stored in an array when it is generated, and it cost time O(log n) to find the proper font in theory, where n is ~= 10. And since the font data is small, (typical 1-2 glyph in it), the speed should be very fast.

I didn't have the comparing data yet, I can do a benchmark if required. But for the current slow problem, it is in the SPI/I2C interface, and the processing code(at least 3 times calling draw functions in the small u8g.firstPage()/u8g.nextPage() loop) is in the big loop which cause the lag.

**Comment by thinkyhead** [comment]
I see. Well, before we could transition to this system, I agree with @AnHardt that the ugly characters and the shifting of the lines needs to be fixed. And of course this PR will need to be rebased. If you need to throw out the changes I made recently in the process, that's okay. But if you can preserve the "smarter" changes that I made, I would appreciate that effort.

**Comment by yhfudev** [comment]
@thinkyhead
I'm not quite sure about the commit f20b2b21eb0434f3812952f6979ed6fac15e0a9f. Is that simplify the call function? to reduce the number of call stack? or remove the return value to "reduce return time"?

In the new language system, I would suggest to use the lcd_print(char*) and lcd_print_wchar(wchar_t) and avoid expose the print interfaces of HD44780 or U8GLIB, so the caller can always assume it called the right API to handle the UTF-8 strings.

I also fix the display problem and the font error.
g-de-1

g-de-2

And the Greek(el) language is also fixed:
g-el

**Comment by yhfudev** [comment]
Can you please make the decision if you want the UTF-8 features integrated to the firmware or not?

**Comment by AnHardt** [comment]
Compared to the normal ASCII symbols all the special symbols (German and Greece) are (still) one pixel too deep.

**Comment by thinkyhead** [comment]
> Can you please make the decision if you want the UTF-8 features integrated to the firmware or not?

@yhfudev Can you please make the decision whether you want to play nicely or not? I can post a link for you on how not to be a jerk to the maintainers, if you need it. Hint 1: don't make petulant demands and please learn to be patient. We all have busy lives here and are doing the best we can.

As for me, I've been quite busy with other projects and don't get much time to review. Sure, maybe I'd like to see this integrated. But I don't have a graphical display currently to test with, and until this gets testing from more users, _of course_ it's not going to get integrated. Only 3 people besides yourself have participated at all in this discussion. That's simply not enough to make a determination whether or not it's ready. And there is still the problem of the special symbols needing to be fixed.

**Comment by thinkyhead** [comment]
@yhfudev I'm working on making a new branch with this feature updated for `RCBugFix`. I'm cherry-picking the changes by hand. I will let you know when it's ready. It will replace this PR.

**Comment by yhfudev** [comment]
@thinkyhead That's a good news!

The issues that might annoy other developers/users which I concerned are:
1. One of U8G C header file should be exposed for including, we need to make sure it's notified in the document. And it would be best if @olikraus be kindly to include the header file in the Arduino path of the releasing package for us so the end users don't need to aware of the changes.
2. The scripts to generate the font header files. I found the font header would not need to be changed after a slight change for most time. I can rewrite the script in python if it helps.

**Comment by thinkyhead** [comment]
@yhfudev I have a branch ready for your review at:
https://github.com/thinkyhead/Marlin/tree/rc_utf8_diff

I'm sure it isn't perfect. We will need to go through it carefully and make sure it works. It's completely built on the latest codebase.

I assume that the `language_data_*.h` files are generated, yes?
If so, I will drop them from the branch. You will probably want to rebuild them.
